File: framework/languages/java/e2e/DiscoveryRegistryHa/Server/Provider/src/main/java/systems/zlink/e2e/discoveryregistryha/provider/ProfileRequestHandler.java
package systems.zlink.e2e.discoveryregistryha.provider;

import systems.zlink.e2e.discoveryregistryha.shared.Contracts;
import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;

@ZLinkHandlerGroup(Contracts.HANDLER_GROUP)
public final class ProfileRequestHandler
    implements ZLinkRequestHandler<Contracts.ProfileRequest, Contracts.ProfileReply> {
    private final ProviderEvidenceStore evidence;

    public ProfileRequestHandler(ProviderEvidenceStore evidence) {
        this.evidence = evidence;
    }

    @Override
    public Contracts.ProfileReply handle(
        Contracts.ProfileRequest request,
        ZLinkRequestContext context) {
        evidence.record(request.marker(), request.value());
        return new Contracts.ProfileReply(
            "profile:" + request.value(),
            evidence.rid(),
            request.marker());
    }
}

File: framework/languages/node/e2e/DiscoveryRegistryHa/Server/Provider/Handlers/profile-request-handler.ts
import { Injectable } from '@nestjs/common';
import type { ZLinkRequestContext, ZLinkRequestHandler } from '@zlink-systems/framework';
import type { ProfileRes, ProfileReq } from '../../../Shared/messages';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@Injectable()
export class ProfileRequestHandler implements ZLinkRequestHandler<ProfileReq, ProfileRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: ProfileReq, context: ZLinkRequestContext): Promise<ProfileRes> {
    const marker = request.marker ?? '';
    this.evidence.add(
      `profile-request|rid=${this.evidence.rid}|value=${request.value}|marker=${marker}|packet=${context.packetName}`
    );
    return { value: `profile:${request.value}`, providerRid: this.evidence.rid, marker: request.marker };
  }
}
